File: 2019/01/problem1.py
# two approaches, in Python

import sys

from itertools import islice, takewhile

from typing import Callable, List, Iterator, TypeVar

_T = TypeVar('_T')

def iterate(fn: Callable[[_T], _T], init: _T) -> Iterator[_T]:
    while True:
        yield init
        init = fn(init)

def fuel(mass: int) -> int:
    return mass // 3 - 2

def total_fuel_proc(mass: int) -> int:
    total = 0
    inc_fuel = fuel(mass)
    while inc_fuel > 0:
        total += inc_fuel 
        inc_fuel = fuel(inc_fuel)
    return total

def total_fuel_func(mass: int) -> int:
    return sum(takewhile(lambda f: f > 0, islice(iterate(fuel, mass), 1, None)))

def main(args: List[str]) -> int:
    if len(args) != 2:
        print(f'Usage: {args[0]} 1|2 <input.txt', file=sys.stderr)
        return 0

    masses = (int(l.rstrip()) for l in sys.stdin.readlines())

    if args[1] == '1':
        print(sum(fuel(m) for m in masses))
    elif args[1] == '2p':
        print(sum(total_fuel_proc(m) for m in masses))
    elif args[1] == '2f':
        print(sum(total_fuel_func(m) for m in masses))
    else:
        print(f'Usage: {args[0]} 1|2p|2f <input.txt', file=sys.stderr)

    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: 2019/01/problem1.rs
use std::{
    env,
    io::{self, BufRead},
    mem,
    time::Instant,
};

struct Iterate<T, F> {
    next: T,
    iter: F,
}

impl<T: Clone, F: FnMut(T) -> T> Iterator for Iterate<T, F> {
    type Item = T;

    fn next(&mut self) -> Option<T> {
        let next = self.next.clone();
        let mut next = (self.iter)(next);
        mem::swap(&mut self.next, &mut next);
        Some(next)
    }
}

fn iterate<T, F>(iter: F, init: T) -> impl Iterator<Item = T>
  where T: Clone, F: FnMut(T) -> T {
    Iterate { next: init, iter }
}

#[inline]
fn fuel(mass: i32) -> i32 {
    mass / 3 - 2
}

#[inline]
fn total_fuel(mass: i32) -> i32 {
    iterate(fuel, mass).skip(1).take_while(|f| *f > 0).sum()
}

#[inline]
fn total_fuel_proc(mass: i32) -> i32 {
    let mut total = 0;
    let mut incr = fuel(mass);
    while incr > 0 {
        total += incr;
        incr = fuel(incr);
    }
    total
}

fn main() {
    if let Some(prob) = env::args().skip(1).next() {
        let stdin = io::stdin();
        let masses = stdin.lock().lines()
            .map(|l| l.unwrap().trim_end().parse::<i32>().unwrap());
        let begin = Instant::now();
        let result = match prob.as_str() {
            "1" => masses.map(fuel).sum::<i32>(),
            "2f" => masses.map(total_fuel).sum::<i32>(),
            "2p" => masses.map(total_fuel_proc).sum::<i32>(),
            _ => {
                eprintln!("Usage: problem1 1|2 <input.txt");
                return;
            },
        };
        let elapsed = begin.elapsed();
        println!("{}", result);
        println!("in {} ms", (elapsed.as_secs() * 1000) as f64 +
            (elapsed.subsec_nanos() as f64 / 1e6));
    } else {
        eprintln!("Usage: problem1 1|2 <input.txt");
    }
}
